pkg/provision/workspace/persistentHome.go

```diff
@@ -0,0 +1,87 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package workspace
+
+import (
+	"fmt"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/constants"
+	"github.com/devfile/devworkspace-operator/pkg/dwerrors"
+)
+
+// Adds a Devfile volume to the given DevWorkspace and mounts the volume to `/home/user/` for every
+// container component defined in the DevWorkspace by adding additional volumeMounts.
+// Has no effect if `persistUserHome` is disabled in the DevWorkspaceOperatorConfig.
+// Returns an error if `persistUserHome` is enabled in the DWOC and at least one of the container components
+// in the DevWorkspace already mounts to `/home/user/`.
+func ProvisionPersistentUserHome(workspace *common.DevWorkspaceWithConfig) error {
+	shouldPersistHome, err := needsPersistentHomeDirectory(workspace)
+	if err != nil {
+		return err
+	}
+	if !shouldPersistHome {
+		return nil
+	}
+
+	homeVolume := v1alpha2.Component{
+		Name: constants.HomeVolumeName,
+		ComponentUnion: v1alpha2.ComponentUnion{
+			Volume: &v1alpha2.VolumeComponent{},
+		},
+	}
+	homeVolumeMount := v1alpha2.VolumeMount{
+		Name: constants.HomeVolumeName,
+		Path: constants.HomeUserDirectory,
+	}
+
+	workspace.Spec.Template.Components = append(workspace.Spec.Template.Components, homeVolume)
+	for _, component := range workspace.Spec.Template.Components {
+		if component.Container == nil {
+			continue
+		}
+		component.Container.VolumeMounts = append(component.Container.VolumeMounts, homeVolumeMount)
+	}
+
+	return nil
+}
+
+// Returns true if `persistUserHome` is enabled in the DevWorkspaceOperatorConfig
+// and none of the container components in the DevWorkspace mount a volume to `/home/user/`.
+// Returns false if `persistUserHome` is disabled.
+// Returns an error if at least one of the DevWorkspace's container components already mount a volume to `/home/user/`.
+func needsPersistentHomeDirectory(workspace *common.DevWorkspaceWithConfig) (bool, error) {
+	if !*workspace.Config.Workspace.PersistUserHome.Enabled {
+		return false, nil
+	}
+	for _, component := range workspace.Spec.Template.Components {
+		if component.Container == nil {
+			continue
+		}
+		for _, volumeMount := range component.Container.VolumeMounts {
+			if volumeMount.Path == constants.HomeUserDirectory {
+				// If a volume is already being mounted to /home/user/, it takes precedence
+				// over the DWO-provisioned home directory volume.
+				return false, &dwerrors.WarningError{
+					Message: fmt.Sprintf("Unable to add persistentHome volume: volume '%s' already mounts to %s", volumeMount.Name, constants.HomeUserDirectory),
+				}
```

Comment:
I wonder if it's worth adding the warning here -- the user is doing nothing wrong and cannot really fix the issue themselves, as it's an administrator-level setting. 

This suggests that folks shouldn't manually configure home directories, which I'm not sure is the case (e.g. if your devfile _needs_ persistent home and you want it to work regardless of whether the config is set).

Perhaps a rewording to be more of an "info" warning than a "warning" warning?

Reply:
> I wonder if it's worth adding the warning here -- the user is doing nothing wrong and cannot really fix the issue themselves, as it's an administrator-level setting. 

Agreed.

I ended up trying to reword the warning, but it might be better to not add a warning to the workspace and instead log this at the controller level (basically, trying to let the admin know what's going on rather than warning the user).

Let me know your thoughts.